cpp/ql/src/Best Practices/Likely Errors/CommaBeforeMisleadingIndentation.ql

```diff
@@ -0,0 +1,66 @@
+/**
+ * @name Comma before misleading indentation
+ * @description The expressions before and after the comma operator can be misread because of an unusual difference in start columns.
+ * @kind problem
+ * @id cpp/comma-before-misleading-indentation
+ * @problem.severity warning
+ * @precision medium
+ * @tags maintainability
+ *       readability
+ */
+
+import cpp
+import semmle.code.cpp.commons.Exclusions
+
+/**
+ * Holds if this is an implicit `this`.
+ *
+ * ThisExpr.isCompilerGenerated() is currently not being extracted, so use a heuristic.
+ */
+predicate isCompilerGenerated(ThisExpr te) {
+  exists(int line, int colStart, int colEnd |
+    te.getLocation().hasLocationInfo(_, line, colStart, line, colEnd)
+  |
+    colStart = colEnd
+    or
+    exists(Call c | c.getQualifier() = te | c.getLocation() = te.getLocation())
+  )
+}
+
+/** Gets the sub-expression of 'e' with the earliest-starting Location */
+Expr normalizeExpr(Expr e) {
+  if forex(Expr q | q = e.(Call).getQualifier() | not isCompilerGenerated(q))
+  then result = normalizeExpr(e.(Call).getQualifier())
+  else
+    if forex(Expr q | q = e.(FieldAccess).getQualifier() | not isCompilerGenerated(q))
+    then result = normalizeExpr(e.(FieldAccess).getQualifier())
+    else
+      if e.hasExplicitConversion()
+      then result = normalizeExpr(e.getFullyConverted())
+      else result = e
+}
+
+predicate isParenthesized(CommaExpr ce) {
+  ce.getParent*().(Expr).isParenthesised()
+  or
+  ce.isUnevaluated() // sizeof(), decltype(), alignof(), noexcept(), typeid()
+  or
+  ce.getParent*() = any(IfStmt i).getCondition()
```

Comment:
Should this predicate also include a `SwitchStmt`'s expression?

Reply:
Mmm, yes, probably.